File: backend/app/schemas/sentiment_eval.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

# 情绪评测只涵盖三分类，和 news_signal_classifier 的输出标签保持一致。
SentimentLabel = Literal["positive", "negative", "neutral"]
SENTIMENT_LABELS: tuple[SentimentLabel, ...] = ("positive", "negative", "neutral")


class SentimentGoldSample(BaseModel):
    """情绪/利好利空金标样本：一段文本 + 人工标注标签，importance 可选。

    title/summary/body/market 为真实新闻采样的可选字段，与线上分类输入对齐；
    缺省时以 text 充当 title（legacy 手写样本行为）。
    """

    sample_id: str
    text: str
    sentiment_label: SentimentLabel
    # 与线上 classify(title, summary, body) 输入对齐的可选字段。
    title: str | None = None
    summary: str | None = None
    body: str | None = None
    # 命中市场（如 us / cn-a / hk），影响中英词典路径。
    market: str | None = None
    # 采样来源 NewsItem id，便于追溯。
    news_id: int | None = None
    # importance 为可选人工标注（0~1 的重要度打分），参与加权准确率计算。
    importance: float | None = Field(default=None, ge=0.0, le=1.0)
    # 模型预测标签，评测前由分类器填充；金标文件里通常为空。
    predicted_sentiment: SentimentLabel | None = None

    @model_validator(mode="after")
    def validate_text(self) -> SentimentGoldSample:
        if not self.text.strip():
            raise ValueError("text is required")
        return self

    @property
    def effective_title(self) -> str:
        return self.title or self.text


class SentimentLabelMetrics(BaseModel):
    label: SentimentLabel
    precision: float = Field(ge=0.0, le=1.0)
    recall: float = Field(ge=0.0, le=1.0)
    f1: float = Field(ge=0.0, le=1.0)
    support: int = Field(ge=0)


class SentimentEvaluationMetrics(BaseModel):
    accuracy: float = Field(ge=0.0, le=1.0)
    macro_f1: float = Field(ge=0.0, le=1.0)
    sample_count: int = Field(ge=0)
    per_label: list[SentimentLabelMetrics]
    # confusion_matrix[actual][predicted] = 计数
    confusion_matrix: dict[str, dict[str, int]]
    # 有 importance 标注的样本按权重的加权准确率（无标注样本权重 1.0）；
    # 数据集内所有样本都没有 importance 标注时为 None。
    importance_weighted_accuracy: float | None = None


class SentimentModelRun(BaseModel):
    model_name: str
    metrics: SentimentEvaluationMetrics


class SentimentLabelDelta(BaseModel):
    label: SentimentLabel
    f1_before: float = Field(ge=0.0, le=1.0)
    f1_after: float = Field(ge=0.0, le=1.0)
    f1_delta: float


class SentimentABComparison(BaseModel):
    model_a: SentimentModelRun
    model_b: SentimentModelRun
    accuracy_delta: float
    macro_f1_delta: float
    label_deltas: list[SentimentLabelDelta]
    winner: Literal["model_a", "model_b", "tie"]
    reason: str

    @model_validator(mode="after")
    def validate_reason(self) -> SentimentABComparison:
        if not self.reason.strip():
            raise ValueError("reason is required")
        return self


class SentimentEvalHistoryEntry(BaseModel):
    """history[].entries[] 的单个模型摘要点位。"""

    model_name: str
    accuracy: float = Field(ge=0.0, le=1.0)
    macro_f1: float = Field(ge=0.0, le=1.0)


class SentimentEvalHistoryPoint(BaseModel):
    """一次 POST /sentiment/run 触发的 batch 摘要（GET 返回最近 20 个）。"""

    batch_id: str
    evaluated_at: datetime
    dataset_hash: str
    sample_count: int
    entries: list[SentimentEvalHistoryEntry]


class SentimentEvalRegression(BaseModel):
    """最新 batch 与上一个同 dataset_hash batch 的同名模型对比（跌幅最大者）。"""

    model_name: str
    previous_macro_f1: float = Field(ge=0.0, le=1.0)
    current_macro_f1: float = Field(ge=0.0, le=1.0)
    delta: float
    regressed: bool


class SentimentEvalResponse(BaseModel):
    """GET /eval/sentiment 的返回体。available=False 表示金标缺失时的降级。"""

    available: bool
    dataset_path: str
    sample_count: int
    primary: SentimentModelRun | None = None
    comparison: SentimentABComparison | None = None
    note: str | None = None
    # 最近一个 batch 的评测时间；库里尚无记录时为 None。
    evaluated_at: datetime | None = None
    # 本 batch 全部模型 run（rule/llm/hybrid），primary 始终指向 rule-baseline。
    runs: list[SentimentModelRun] = Field(default_factory=list)
    # 是否存在激活的 LLM provider 配置（与本 batch 运行时是否用了 LLM 无关，取当前状态）。
    llm_available: bool = False
    # 最近 20 个 batch 的 macro_f1 走势摘要。
    history: list[SentimentEvalHistoryPoint] = Field(default_factory=list)
    # 与上一个同 dataset_hash batch 的回归对比；无可比较历史时为 None。
    regression: SentimentEvalRegression | None = None

File: frontend/src/api/mock/sentimentEval.ts
// 情绪/利好利空评测 (Sentiment Eval Harness) mock 数据：GET /api/eval/sentiment 与
// POST /api/eval/sentiment/run 共用同一份夹具（重新评测在离线模式下返回同一快照）。
//
// 覆盖设计文档 docs/superpowers/specs/2026-08-02-sentiment-eval-revamp-design.md
// 的三个模型 run（rule-baseline / llm / hybrid）、history 走势与 regression 回归对比，
// 便于在后端工作块 B 落地前离线联调前端展示。

import type { SentimentEvalResponse } from '../../types/api';
import { isoMinutesAgo } from './shared';

function metrics(accuracy: number, macroF1: number, importanceWeighted: number | null) {
  return {
    accuracy,
    macro_f1: macroF1,
    sample_count: 120,
    per_label: [
      { label: 'positive' as const, precision: accuracy + 0.02, recall: accuracy - 0.03, f1: macroF1 + 0.01, support: 40 },
      { label: 'negative' as const, precision: accuracy - 0.04, recall: accuracy - 0.01, f1: macroF1 - 0.02, support: 40 },
      { label: 'neutral' as const, precision: accuracy - 0.01, recall: accuracy + 0.02, f1: macroF1 + 0.01, support: 40 },
    ],
    confusion_matrix: {
      positive: { positive: 32, negative: 4, neutral: 4 },
      negative: { positive: 3, negative: 30, neutral: 7 },
      neutral: { positive: 2, negative: 5, neutral: 33 },
    },
    importance_weighted_accuracy: importanceWeighted,
  };
}

const ruleRun = { model_name: 'rule-baseline', metrics: metrics(0.7, 0.68, 0.71) };
const llmRun = { model_name: 'llm:deepseek/deepseek-chat', metrics: metrics(0.85, 0.82, 0.86) };
const hybridRun = { model_name: 'hybrid:deepseek/deepseek-chat', metrics: metrics(0.83, 0.8, 0.84) };

export const mockSentimentEval: SentimentEvalResponse = {
  available: true,
  dataset_path: 'data/gold/sentiment_v2.jsonl',
  sample_count: 120,
  primary: ruleRun,
  runs: [ruleRun, llmRun, hybridRun],
  llm_available: true,
  comparison: {
    model_a: ruleRun,
    model_b: llmRun,
    accuracy_delta: 0.15,
    macro_f1_delta: 0.14,
    label_deltas: [
      { label: 'positive', f1_before: 0.69, f1_after: 0.83, f1_delta: 0.14 },
      { label: 'negative', f1_before: 0.66, f1_after: 0.8, f1_delta: 0.14 },
      { label: 'neutral', f1_before: 0.69, f1_after: 0.83, f1_delta: 0.14 },
    ],
    winner: 'model_b',
    reason: 'llm:deepseek/deepseek-chat 在全部标签上均领先 rule-baseline。',
  },
  note: null,
  evaluated_at: isoMinutesAgo(12),
  history: [
    {
      batch_id: 'batch-mock-0003',
      evaluated_at: isoMinutesAgo(12),
      dataset_hash: 'ab12cd34ef56ab78',
      sample_count: 120,
      entries: [
        { model_name: 'rule-baseline', accuracy: 0.7, macro_f1: 0.68 },
        { model_name: 'llm:deepseek/deepseek-chat', accuracy: 0.85, macro_f1: 0.82 },
        { model_name: 'hybrid:deepseek/deepseek-chat', accuracy: 0.83, macro_f1: 0.8 },
      ],
    },
    {
      batch_id: 'batch-mock-0002',
      evaluated_at: isoMinutesAgo(1500),
      dataset_hash: 'ab12cd34ef56ab78',
      sample_count: 120,
      entries: [
        { model_name: 'rule-baseline', accuracy: 0.69, macro_f1: 0.67 },
        { model_name: 'llm:deepseek/deepseek-chat', accuracy: 0.88, macro_f1: 0.86 },
        { model_name: 'hybrid:deepseek/deepseek-chat', accuracy: 0.84, macro_f1: 0.81 },
      ],
    },
    {
      batch_id: 'batch-mock-0001',
      evaluated_at: isoMinutesAgo(4200),
      dataset_hash: '9f8e7d6c5b4a3210',
      sample_count: 96,
      entries: [
        { model_name: 'rule-baseline', accuracy: 0.66, macro_f1: 0.64 },
        { model_name: 'llm:deepseek/deepseek-chat', accuracy: 0.8, macro_f1: 0.77 },
      ],
    },
  ],
  regression: {
    model_name: 'llm:deepseek/deepseek-chat',
    previous_macro_f1: 0.86,
    current_macro_f1: 0.82,
    delta: -0.04,
    regressed: true,
  },
};

/** 未配置 LLM 时的降级夹具：仅规则阈值 A/B（legacy 行为），无 runs 中的 llm/hybrid。 */
export const mockSentimentEvalNoLlm: SentimentEvalResponse = {
  available: true,
  dataset_path: 'data/gold/sentiment_v2.jsonl',
  sample_count: 120,
  primary: ruleRun,
  runs: [ruleRun],
  llm_available: false,
  comparison: {
    model_a: ruleRun,
    model_b: { model_name: 'rule-sensitive (±0.10)', metrics: metrics(0.72, 0.7, 0.73) },
    accuracy_delta: 0.02,
    macro_f1_delta: 0.02,
    label_deltas: [
      { label: 'positive', f1_before: 0.69, f1_after: 0.71, f1_delta: 0.02 },
      { label: 'negative', f1_before: 0.66, f1_after: 0.68, f1_delta: 0.02 },
      { label: 'neutral', f1_before: 0.69, f1_after: 0.71, f1_delta: 0.02 },
    ],
    winner: 'model_b',
    reason: '未配置 LLM，仅规则阈值对比。',
  },
  note: '未配置 LLM，仅规则阈值对比。',
  evaluated_at: isoMinutesAgo(30),
  history: [],
  regression: null,
};

/** 金标存在但库里无记录时的引导空状态夹具。 */
export const mockSentimentEvalNoRuns: SentimentEvalResponse = {
  available: true,
  dataset_path: 'data/gold/sentiment_v2.jsonl',
  sample_count: 120,
  primary: null,
  runs: [],
  llm_available: true,
  comparison: null,
  note: '尚无评测记录，请点击重新评测。',
  evaluated_at: null,
  history: [],
  regression: null,
};
